Build-process.md

```diff
@@ -0,0 +1,114 @@
+
+* `DEV_MODE`: Set to `true` if and only if we are in dev mode. Otherwise, this is `false`.
+* `EMULATOR_MODE`: Set to `true` if and only if we are not in deploy mode. Note that being in deploy mode implies being in prod mode. Otherwise, this is `false`.
+* `ENABLE_MAINTENANCE_MODE`: Set to `true` if and only if we are in maintenance mode. Note that being in maintenance mode implies being in prod mode. Otherwise, this is `false`.
+
+### Dev mode
+
+We use dev mode for local development. For example, when you run `python -m scripts.start`, you build the app in dev mode. This mode is optimized to build as quickly as possible, so it doesn't do much optimization.
+
+Here's what happens during a dev mode build:
+
+1. Some third-party libraries are generated. If you recently ran the server, then there might not be much to change here. Otherwise, the third-party libraries from `manifest.json` are downloaded. Note that the libraries in `manifest.json` are just some of our frontend libraries. The downloaded CSS and JavaScript code files are combined into `third_party/generated/third_party.css` and `third_party/generated/third_party.js`, and the downloaded fonts are placed in `third_party/webfonts/`.
+2. The build script sets the constants in `assets/constants.ts` to enable dev mode. Specifically, `DEV_MODE` is `true`, `EMULATOR_MODE` is `true`, and the other constants are `false`.
+3. [Webpack](https://webpack.js.org/) bundles our frontend code files and runs with a [dev configuration](https://github.com/oppia/oppia/blob/develop/webpack.dev.config.ts) designed for minimal compilation time. Note that the dev mode configuration file inherits from the [root webpack config file](https://github.com/oppia/oppia/blob/develop/webpack.common.config.ts). In dev mode, webpack runs in "watch" mode so that when files are changed, webpack re-builds the app automatically. Note that **this step is not handled by `build.py`.** Instead, this step is usually performed by whatever script you actually executed, for example `start.py`. When running the frontend tests, this step is triggered by configuration in `karma.conf.ts`.
+4. The `app_dev.yaml` configuration file for the app already exists, so it doesn't need to be generated.
+
+### Prod mode
+
+In prod mode, we more aggressively optimize the app for performance, but the cost of this optimization is that the build process takes longer. Prod mode is also a little more complicated because it has three additional options: minify only third-party libraries, maintenance mode, and deploy mode.
+
+#### Plain prod mode
+
+First, let's consider what happens in plain prod mode where those three options are all disabled:
+
+1. The third-party libraries are generated like in [dev mode](#dev-mode), but they are also minified so they take up less space.
+2. Hashes are generated for asset files like images. These will be used later.
```

Comment:
Maybe change to `later (see step 5 below)`.

Reply:
done